### PR TITLE
Bump doorkeeper from 5.0.2 to 5.3.1

### PR DESCRIPTION
**Pull request by edwardkerry**
This PR bumps Doorkeeper from version 5.0.2 to 5.3.1, through the following versions.

[5.1.0 rc1](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#510rc1)
[5.1.0 rc2](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#510rc2)
[5.1.0](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#510)
[5.2.0 rc1](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#520rc1)
[5.2.0 rc2](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#520rc2)
[5.2.0 rc3](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#520rc3)
[5.2.0](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#520)
[5.2.1](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#521)
[5.2.2](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#522)
[5.2.3](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#523)
[5.2.4](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#524)
[5.3.0](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#530)
[5.3.01](https://github.com/doorkeeper-gem/doorkeeper/blob/master/CHANGELOG.md#531)

5.2.0rc3 introduced some breaking changes for Signon, particularly around the SigninRequiredAuthorizationsController (which is a customisation of Doorkeeper's AuthorizationController),  as [this PR](https://github.com/doorkeeper-gem/doorkeeper/pull/1277) made the server setting a default `scope` mandatory. See individual commit messages for more details.

[trello](https://trello.com/c/NWjFMEnl/1767-5-upgrade-doorkeeper-gem-in-signon)

### CODE REVIEW

File: config/initializers/doorkeeper.rb [hunk 1]
```diff
@@ -16,32 +21,389 @@
...
   # Define access token scopes for your provider
-  # For more information go to https://github.com/applicake/doorkeeper/wiki/Using-Scopes
-  # authorization_scopes do
-  #   scope :public, :default => true, :description => "The public one"
-  #   scope :write,  :description => "Updating information"
+  # For more information go to
+  # https://doorkeeper.gitbook.io/guides/ruby-on-rails/scopes
+  #
+  # default_scopes  :public
+  # optional_scopes :write, :update
```
Comment (cbaines) [inline config/initializers/doorkeeper.rb]: Do these bits of configuration relate to the same "scopes" that the monkey patching involves? https://github.com/alphagov/signon/pull/1358/commits/08fea4cc9bf125bf1ebd8a8b63e73732485117b2

File: config/initializers/doorkeeper.rb [hunk 2]
```diff
@@ -16,32 +21,389 @@
...
+  #
+  # allow_token_introspection false
+  #
+  # If you need to block the request at all, then configure your routes.rb or web-server
+  # like nginx to forbid the request.
+
+  # WWW-Authenticate Realm (default: "Doorkeeper").
+  #
+  # realm "Doorkeeper"
+end
```
Comment (cbaines) [inline config/initializers/doorkeeper.rb]: The newline at the end is missing.

File: config/initializers/doorkeeper_scopes.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+  def validate_scopes
+    return true if scope.blank?
+    old_validate_scopes
+  end
+
+  def validate_params
+    response_type.blank? ? :response_type : true
+  end
+end
```
Comment (cbaines) [inline config/initializers/doorkeeper_scopes.rb]: How does this change here relate to https://github.com/alphagov/signon/pull/1358/commits/30c5f2b1dbc19e34e620399edfa8215bc4970da6 ? Seems to me that they might be related, but I'm unsure?

**Comment by cbaines** [comment]
Looks generally good to me :+1: I've asked a few rather uninformed questions unfortunately, mostly just so I can try and understand the changes better.

File: config/initializers/doorkeeper_scopes.rb (hunk 3 above)
Comment (edwardkerry) [reply config/initializers/doorkeeper_scopes.rb]: It's replacing this method which rejects a request if scope params are blank
https://github.com/doorkeeper-gem/doorkeeper/blob/5795bc26627d3899ae813ecb5fa64289f5345e45/lib/doorkeeper/oauth/pre_authorization.rb#L95

As https://github.com/alphagov/signon/pull/1358/commits/30c5f2b1dbc19e34e620399edfa8215bc4970da6 sets all `scope`s  as empty strings, we essentially want to ensure that a blank `scope` param is still valid.

File: config/initializers/doorkeeper.rb (hunk 1 above)
Comment (edwardkerry) [reply config/initializers/doorkeeper.rb]: It does, if using scopes is possible to set a default scope which will be used if none is explicitly set, and other options to allow or deny various auth attempts. 

As we are not using scopes, the data migration sets all scopes to `''` and other changes make this a valid option.

Comment (cbaines) [reply config/initializers/doorkeeper.rb]: Ok, thanks.

**Comment by edwardkerry** [comment]
> I think it would be good to soon look again at Doorkeeper, and try and bring our use of it more in line with the OAuth standard, as well as trying to reduce the monkey patching of the internals, but I don't think addressing these two issues is enough to not merge these changes.

Agreed, I will write up a card for us to pick this up soon.

**Pull request merged**